kong/tools/wrpc/message.lua

```diff
@@ -0,0 +1,174 @@
+
+local pb = require "pb"
+
+local tonumber = tonumber
+local table_unpack = table.unpack -- luacheck: ignore
+local select = select
+
+local pb_decode = pb.decode
+local pb_encode = pb.encode
+
+local ngx_log = ngx.log
+local ERR = ngx.ERR
+local NOTICE = ngx.NOTICE
+local ngx_now = ngx.now
+
+-- utility functions
+
+--- little helper to ease grabbing an unspecified number
+--- of values after an `ok` flag
+local function ok_wrapper(ok, ...)
+  return ok, {n = select('#', ...), ...}
+end
+
+local empty = {}
+
+--- decodes each element of an array with the same type
+local function decodearray(decode, typ, l)
+  local out = {}
+  -- protobuf encodes {} as the same of nil so we have to do this.
+  for i, v in ipairs(l or empty) do
+    out[i] = decode(typ, v)
+  end
+  return out
+end
+
+--- encodes each element of an array with the same type
+local function encodearray(encode, typ, l)
+  local out = {}
+  for i = 1, l.n do
+    out[i] = encode(typ, l[i])
+  end
+  return out
+end
+
+
+local _M = {}
+
+local function send_error(wrpc_peer, payload, error)
+  return wrpc_peer:send_payload({
+    mtype = "MESSAGE_TYPE_ERROR",
+    error = error,
+    svc_id = payload.svc_id,
+    rpc_id = payload.rpc_id,
+    ack = payload.seq,
+  })
+end
+
+--- Handle RPC data (mtype == MESSAGE_TYPE_RPC).
+--- Could be an incoming method call or the response to a previous one.
+--- @param payload table decoded payload field from incoming `wrpc.WebsocketPayload`
+function _M.process_message(wrpc_peer, payload)
+  local rpc = wrpc_peer.service:get_rpc(
+    payload.svc_id .. '.' .. payload.rpc_id)
+  if not rpc then
+    send_error(wrpc_peer, {
```

Comment:
the `error` parameter seems not set in the `send_error` function. we should also check return of this function. consider doing something like this:
```lua
return send_error(wrpc_peer, payload, {
      etype = "ERROR_TYPE_INVALID_SERVICE",
      description = "Invalid service (or rpc)",
    })
```
then in send_error, do
```lua
local function send_error(wrpc_peer, payload, error)
  local ok, err = wrpc_peer:send_payload({
    mtype = "MESSAGE_TYPE_ERROR",
    error = error,
    svc_id = payload.svc_id,
    rpc_id = payload.rpc_id,
    ack = payload.seq,
  })
  if not ok then
    return ok, err
  end
  return nil, err.description
end
```

Reply:
I'm missing the payload argument.
Suggestion applied.